Add CONTRIBUTING.md (#762)
rmccue/Requests: .github/CONTRIBUTING.md

## Process

1. Ideally, start with [opening an issue][new-issue] to check the need for a PR. It's possible that a feature may be rejected at an early stage, and it's better to find out before you write the code.
    Note: There may be an issue or PR open already. If so, please join the discussion in that issue or PR instead of opening a duplicate issue/PR.
2. Fork [the repository][] on Github, create a new branch off the `develop` branch and write the code.
    - Small, atomic commits are preferred.
    - Explain the motivation behind the change in the commit message and if relevant, provide links to external sources on which you based your change.
    - Please ensure any new features and bug fixes are accompanied by tests which fully cover the change.
3. File a PR and fill out the PR template. If the PR isn't ready for review/merge yet, open the PR [as a draft][draft-prs]. If your PR closes an existing issue, add "Fixes #xxx" to the message, so that the issue will be closed when the PR is merged.
4. If needed, iterate on the code until it is ready. When you're ready, mark the PR as "ready for review".

> 5. Automated checks will be run on all PRs via [GitHub Actions][[ghactions]]. Please fix any issues identified by these automated checks promptly. It is unlikely your PR will be reviewed until the automated checks pass.

Reviewer: ```suggestion
5. Automated checks will be run on all PRs via [GitHub Actions][ghactions]. Please fix any issues identified by these automated checks promptly. It is unlikely your PR will be reviewed until the automated checks pass.
```

6. A committer will review your code and offer you feedback.

> 7. Update with the feedback as necessary.

Reviewer: Should we point to the fact that this takes you back to step 5. here, and not automatically gets your PR merged?

8. PR will be merged.

Note:

* All code needs to go through peer review. Committers may not merge their own PR.

If you have questions while working on your contribution and you use Slack, there is
a [#core-http-api][] channel available in the [WordPress Slack][] in which contributions can be discussed.

[new-issue]: https://github.com/WordPress/Requests/issues/new/choose
[the repository]: https://github.com/WordPress/Requests

> [draft-prs]: https://github.blog/2019-02-14-introducing-draft-pull-requests/

Reviewer: Wouldn't it be preferable to link to the documentation instead of a marketing post?
```suggestion
[draft-prs]: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/about-pull-requests#draft-pull-requests
```

Author: Docs links seem to keep changing at GH, so let's just use that intro blog post for now as it contains links to the docs (which GH will hopefully update when needed) anyhow.

Tests are organized as follows:
* Each class in the `src` directory has a corresponding subdirectory in the `tests` directory.
    Example: the tests for the `\WpOrg\Requests\Cookie` class in `src/Cookie.php` can be found in the `tests/Cookie/` subdirectory.
* Unit tests for a specific method will be in this subdirectory in a file/class matching the method name.
    Example: unit tests for the `\WpOrg\Requests\Cookie::domain_matches()` method are in the `WpOrg\Requests\Tests\Cookie\DomainMatchesTest` class in the `tests/Cookie/DomainMatchesTest.php` file.
* Integration tests for a class will be in this subdirectory in a file/class matching the class name.
    Example: integration tests for the `\WpOrg\Requests\Cookie` class are in the `WpOrg\Requests\Tests\Cookie\CookieTest` class in the `tests/Cookie/CookieTest.php` file.

If at all possible, features **must** be unit tested.
All tests should have `@covers` tags pointing to the specific code which is being tested.
Code coverage is monitored for every PR and for the code base as a whole using [CodeCov][].

> [PHPUnit]: https://phpunit.readthedocs.io/en/stable/

Reviewer: The `stable` path unfortunately doesn't seem to work and produces a broken link.
Should we link to 9.5 instead? This would be the version that we are writing the tests against with the polyfills. Updating the polyfills would then mean checking the docs link here if it needs updating...
```suggestion
[PHPUnit]: https://phpunit.readthedocs.io/en/9.5/
```

Author: Updated to the new canonical `main`.

[PHPUnit Polyfills]: https://github.com/Yoast/PHPUnit-Polyfills/
[CodeCov]: https://app.codecov.io/gh/WordPress/Requests/branch/develop

### Prerequisites for running the tests

- [PHP][] >= 5.6
- [Composer][]
- [Python 3][]
- [pip3][]

> - [mitmproxy][] (`pip3 install mitmproxy`)

Reviewer: Python 3 and `pip3` are technically not dependencies of this project, they are just a means to install `mitmproxy`. If you manage to install it in any other way, I don't think we need Python and `pip` as dependencies, do we?

As an example, you can easily build an executable Docker container called `mitmproxy` that bundles Python 3 so you don't need to install it on your host environment...

Reviewer: Which makes me think we should provide something like that to simplify onboarding at some point.

Author: This is not a list of dependencies for the project, but a list of the "Prerequisites for running the tests"....

Author: Removing pip and leaving the rest for the "_Which makes me think we should provide something like that to simplify onboarding at some point._" action ;-)

[PHP]: https://www.php.net/
[Composer]: http://getcomposer.org/
[Python 3]: https://www.python.org/
[pip3]: https://pypi.org/project/pip/
[mitmproxy]: https://mitmproxy.org/

### Running the Tests

```bash
# Start the test server
PORT=8080 vendor/bin/start.sh
export "REQUESTS_TEST_HOST_HTTP=localhost:8080"

# Start the proxy server
PORT=9002 tests/utils/proxy/start.sh
PORT=9003 AUTH="test:pass" tests/utils/proxy/start.sh
export "REQUESTS_HTTP_PROXY=localhost:9002"
export "REQUESTS_HTTP_PROXY_AUTH=localhost:9003"
export "REQUESTS_HTTP_PROXY_AUTH_USER=test"
export "REQUESTS_HTTP_PROXY_AUTH_PASS=pass"

# Run the tests
composer test

# Stop the proxy server
PORT=9002 tests/utils/proxy/stop.sh
PORT=9003 tests/utils/proxy/stop.sh

# Stop the test server
vendor/bin/stop.sh
```

To run the test with code coverage, use `composer coverage` instead.

## Website

A PR to update the [website][] containing the documentation for Requests is automatically generated via GitHub Actions whenever a new release is tagged.

> The website is a [GitHub Pages][] website, build up from the following sources:

Reviewer: ```suggestion
The website is a [GitHub Pages][] website, built up from the following sources:
```

* Templates, styling and a few markdown files in the `gh-pages` branch.

> * The README file and the prose documentation in the `docs` directory of the `stable` branch, which are automatically converted for use in GH Pages using the scripts in the [`build/ghpages`][] directory.

Reviewer: ```suggestion
* The `README` file and the prose documentation in the `docs` directory of the `stable` branch, which are automatically converted for use in GH Pages using the scripts in the [`build/ghpages`][] directory.
```

* API documentation, which is based on the docblocks in the source code, generated using [phpDocumentor][].

[website]: https://requests.ryanmccue.info/
[GitHub Pages]: https://docs.github.com/en/pages
[`build/ghpages`]: https://github.com/WordPress/Requests/blob/develop/build/ghpages/README.md
[phpDocumentor]: https://www.phpdoc.org/

## Releases

> There is [release checklist][] available detailing the release workflow.

Reviewer: ```suggestion
There is a [release checklist][] available detailing the release workflow.
```
